src/packet.ml

```diff
@@ -326,46 +287,33 @@ module Tls_crypt = struct
   let clear_hdr_len =
     hdr_len hmac_len - 1 (* not including acked sequence numbers *)
 
-  let to_be_signed_header ?(more = 0) op header =
-    (* op_key ++ session_id ++ replay_id ++ timestamp ++ ack_len ++ acks ++ remote_session ++ sequence_number *)
-    let acks_len = List.length header.ack_sequence_numbers * id_len
-    and rses_len = if Option.is_some header.remote_session then 8 else 0 in
-    let buflen =
-      1 (* operation *) + session_id_len
-      + id_len + 4 (* timestamp *) + 1
-      (* length of ack list *) + acks_len
-      + rses_len + more
+  let to_be_signed op key header decrypted =
+    let open Result.Syntax in
+    let+ bytes_to_use =
+      match op with
+      | Hard_reset_client_v3 ->
+          (* wkc is not part of the signed *)
+          let* () = guard (Cstruct.length decrypted > 1) `Partial in
+          let arr_len = Cstruct.get_uint8 decrypted 0 in
+          let byte_len =
+            1 (* arr_len *) + (arr_len * id_len)
+            (* acks *) + (if arr_len = 0 then 0 else 8)
+            (* remote session *) + 4 (* sequence number *)
+          in
+          let+ () = guard (Cstruct.length decrypted > byte_len) `Partial in
+          byte_len
+      | _ -> Ok (Cstruct.length decrypted)
     in
-    let buf = Cstruct.create buflen in
-    Cstruct.set_uint8 buf 0 op;
+    let hdr_len = hdr_len 0 in
+    let len = hdr_len + bytes_to_use in
+    let buf = Cstruct.create len in
+    Cstruct.set_uint8 buf 0 (op_key op key);
     Cstruct.BE.set_uint64 buf 1 header.local_session;
     Cstruct.BE.set_uint32 buf 9 header.replay_id;
     Cstruct.BE.set_uint32 buf 13 header.timestamp;
-    Cstruct.set_uint8 buf 17 (List.length header.ack_sequence_numbers);
-    let enc_ack idx ack = Cstruct.BE.set_uint32 buf (18 + (4 * idx)) ack in
-    List.iteri enc_ack header.ack_sequence_numbers;
-    Option.iter
-      (Cstruct.BE.set_uint64 buf (18 + acks_len))
-      header.remote_session;
-    (buf, 18 + acks_len + rses_len)
-
-  let to_be_signed_control op (header, sequence_number, payload) =
-    let buf, off =
-      to_be_signed_header op header ~more:(id_len + Cstruct.length payload)
-    in
-    Cstruct.BE.set_uint32 buf off sequence_number;
-    Cstruct.blit payload 0 buf (off + id_len) (Cstruct.length payload);
+    Cstruct.blit decrypted 0 buf hdr_len bytes_to_use;
     buf
```

Comment:
Hmmm hmm. There is potentially some payload as well that we should cover in our hmac. I think it is better if we cut out the wkc earlier. Probably our hard reset v3 parsing is not optimal?

Reply:
I'd like - with this PR - to do as little parsing work before validating the packet..

Reply:
Right, there ***shouldn't** be any payload in a hard_reset_client_v3 packet (as of the current protocol). The server would need to parse the wkc in the hard_reset_client_v3 in any case in order to get the unwrapped key so it's not any additional parsing.